File: app/src/main/java/com/example/mvvmtest/view/activity/ChatActivity.java
package com.example.mvvmtest.view.activity;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.text.TextUtils;
import android.widget.Button;
import android.widget.EditText;

import com.example.mvvmtest.R;
import com.example.mvvmtest.adapter.ChatAdapter;
import com.example.mvvmtest.dagger.component.ApiController;
import com.example.mvvmtest.model.Message;
import com.example.mvvmtest.viewmodel.ChatViewModel;
import com.github.nkzawa.socketio.client.Socket;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

public class ChatActivity extends AppCompatActivity {

    private ChatViewModel chatViewModel;
    private ChatAdapter chatAdapter;

    @BindView(R.id.chatRecyclerView)
    protected RecyclerView recyclerView;

    @BindView(R.id.sendButton)
    protected Button sendButton;

    @BindView(R.id.sendEditText)
    protected EditText sendEditText;

    @BindView(R.id.toolbar)
    protected Toolbar toolbar;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_chat);

        ButterKnife.bind(this);
        ApiController.getAppComponent().inject(this);
        initViewModel();
        initAdapter();
        checkChatStatus();

    }

    private void checkChatStatus() {
        if (chatViewModel.isConnected()) {
            toolbar.setSubtitle("Connected");
        } else {
            toolbar.setSubtitle("Connected");
        }
    }

    private void initAdapter() {
        chatAdapter = new ChatAdapter(ChatActivity.this, chatViewModel.getMessages().getValue());
        recyclerView.setAdapter(chatAdapter);
        recyclerView.setLayoutManager(new LinearLayoutManager(ChatActivity.this));
    }

    private void initViewModel() {
        chatViewModel = ViewModelProviders.of(this).get(ChatViewModel.class);
        chatViewModel.init();
        chatViewModel.getMessages().observe(this, new Observer<List<Message>>() {
            @Override
            public void onChanged(@Nullable List<Message> messages) {
                chatAdapter.notifyDataSetChanged();
            }
        });

    }

    @OnClick(R.id.sendButton)
    protected void sendMessage() {
        if (TextUtils.isEmpty(sendEditText.getText().toString().trim())) {
            return;
        }
        sendEditText.setText("");
        chatViewModel.sendMessage(new Message());
    }


}

File: app/src/main/java/com/example/chatapp/activity/ChatActivity.kt
package com.example.chatapp.activity

import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager

import android.os.Bundle
import android.text.TextUtils
import android.util.Log
import android.view.View

import com.example.chatapp.R
import com.example.chatapp.adapter.ChatAdapter
import com.example.chatapp.dagger.component.ApiController
import com.example.chatapp.manager.FlechPreferences
import com.example.chatapp.model.Message
import com.example.chatapp.model.Typing
import com.example.chatapp.util.*
import com.example.chatapp.viewmodel.ChatViewModel
import com.github.nkzawa.socketio.client.Socket

import javax.inject.Inject

import kotlinx.android.synthetic.main.content_chat.*

class ChatActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var chatViewModel: ChatViewModel
    private lateinit var chatAdapter: ChatAdapter
    private lateinit var layoutManager: LinearLayoutManager
    private var nickname: Int? = null

    @Inject
    lateinit var socket: Socket

    @Inject
    lateinit var flechPreferences: FlechPreferences

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_chat)
        ApiController.getAppComponent().inject(this)
        //nickname = intent.getIntExtra("nickname", Constant.NICKNAME)
        nickname = Constant.NICKNAME
        setupViews()
        initViewModel()
        initAdapter()
        initViews()
        initSocket()
        setupEditTextBehavior()

    }

    fun setupViews() {

    }

    private fun setupEditTextBehavior() {
        sendEditText.afterTextChanged {
            if (it.isEmpty()) sendTyping(false) else sendTyping(true)
        }
    }

    private fun initViews() {
        sendButton.setOnClickListener(this)
    }

    private fun initAdapter() {
        layoutManager = LinearLayoutManager(this@ChatActivity)
        layoutManager.stackFromEnd = true
        chatRecyclerView.layoutManager = layoutManager
        val mutableList: MutableList<Message> = arrayListOf()
        chatAdapter = ChatAdapter(mutableList)
        chatRecyclerView.adapter = chatAdapter

        chatRecyclerView.onLayoutChange {
            if (it) scrollToBottom()
        }

    }

    private fun scrollToBottom() {
        layoutManager.smoothScrollToPosition(chatRecyclerView, null, chatAdapter.itemCount)
    }

    private fun initViewModel() {

        val viewModelFactory = ChatViewModelFactory(nickname!!)
        chatViewModel = ViewModelProviders.of(this@ChatActivity, viewModelFactory)
                .get(ChatViewModel::class.java)

        chatViewModel.getMessages()
        chatViewModel.messagesLiveData.observe(this, Observer {
            if (it != null) {
                chatAdapter.addMessages(it)
                scrollToBottom()
            }
        })

        chatViewModel.typingLiveData.observe(this, Observer {
            performTyping(it)
        })

    }

    private fun performTyping(typing: Boolean) {
        Log.d(Constant.TAG, "entro en performTyping $typing")
        if (typing) chatLinearLayout.visibility = View.VISIBLE else chatLinearLayout.visibility = View.GONE
    }

    private fun initSocket() {
        socket.connect()
        socket.on(Constant.SocketEvent.NEW_MESSAGE, chatViewModel.onNewMessageListener)
        socket.on(Constant.SocketEvent.TYPING, chatViewModel.onTypingListener)

    }

    private fun sendMessage() {
        sendTyping(false)
        val text = sendEditText!!.text.toString().trim { it <= ' ' }
        if (TextUtils.isEmpty(text)) {
            return
        }
        chatViewModel.sendMessage(socket, Message(Constant.SENDER, nickname!!, text))
        sendEditText.setText("")
    }

    private fun sendTyping(typing: Boolean) {
        chatViewModel.sendTyping(socket, Typing(Constant.SENDER, Constant.NICKNAME, typing))
    }

    private fun disconnect() {
        socket.off(Constant.SocketEvent.NEW_MESSAGE)
        socket.off(Constant.SocketEvent.TYPING)
        socket.disconnect()
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.sendButton -> {
                sendMessage()
            }
            else -> {

            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        disconnect()
    }

}
